packages/next-auth/src/nextAuthPlugin.ts

```diff
@@ -3,39 +3,71 @@ import {
   PartialStlContext,
   MakeStainlessPlugin,
   Params,
+  StlContext,
 } from "stainless";
 import { IncomingMessage, ServerResponse } from "http";
 import { NextServerContext } from "@stl-api/next";
 import { AuthOptions, Session, getServerSession } from "next-auth";
 
+/// An authenticated session, with user data available.
+export type UserSession = Session & {
+  user: {
+    name?: string | null
+    email?: string | null
+    image?: string | null
+  }
+}
+
 declare module "stainless" {
   interface StlContext<EC extends AnyEndpoint> {
-    session?: Session | null;
+    /// If a route is authenticated, handlers will always have access to a valid user session.
+    /// Otherwise, a session is provided if a user is logged in.
+    session: EC['config'] extends { authenticated: true } ? UserSession : UserSession | undefined;
+  }
+
+  interface EndpointConfig {
+    authenticated?: boolean,
   }
 }
 
+function isUserSession(session: Session): session is UserSession {
+  return !!session.user;
+}
+
 export const makeNextAuthPlugin =
   ({ authOptions }: { authOptions: AuthOptions }): MakeStainlessPlugin<any> =>
-  (stl) => ({
-    async middleware<EC extends AnyEndpoint>(
-      endpoint: EC,
-      params: Params,
-      context: PartialStlContext<any, EC>
-    ) {
-      const {
-        args: [req, res],
-      } = requireNextServerContext(context);
-      // TODO catch invalid credentials errors,
-      // send appropriate error response,
-      // and somehow signal to stl.execute to early
-      // exit
-      if (req instanceof IncomingMessage && res instanceof ServerResponse) {
-        context.session = await getServerSession(req, res, authOptions);
-      } else {
-        context.session = await getServerSession(authOptions);
-      }
-    },
-  });
+    (stl) => ({
+      async middleware<EC extends AnyEndpoint>(
+        endpoint: EC,
+        params: Params,
+        context: PartialStlContext<any, EC>
+      ) {
+        const {
+          args: [req, res],
+        } = requireNextServerContext(context);
+
+        // TODO catch invalid credentials errors,
+        // send appropriate error response,
+        // and somehow signal to stl.execute to early
+        // exit
+
+        let session: Session | null | undefined;
```

Comment:
why both null and undefined as options? Is there a meaningful distinction?

Reply:
NextJS represents no session via `null`. Right now I'm choosing to represent that as `undefined` instead as the conditional context type stuff we're doing is pretty finnicky, and I haven't been able to get correct type-checking without casting or expecting errors with `null`. Might be worth doing that to get a clearer API though.

Reply:
okay yeah we can pair on this (or you can pair with andy) to figure out how to make it work with the types